# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [25]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [26]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [27]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [28]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [29]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [30]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [31]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [32]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [33]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [34]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [35]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [36]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [37]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [38]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [39]:
data.duplicated().sum()

71

In [40]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [41]:
import pandas as pd

def categorize_income(income):
    if income <= 30000:
        return 'E'
    elif income <= 50000:
        return 'D'
    elif income <= 200000:
        return 'C'
    elif income <= 1000000:
        return 'B'
    else:
        return 'A'


In [49]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [50]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [44]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [45]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [46]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')
# Удаление строк с аномальными значениями в столбце 'children'.
data = data[(data['children'] != -1) & (data['children'] != 20)]
# Произвели группировку данных по категории "количество детей"
grouped = data.groupby('children')['debt'].agg(['count', 'sum'])
# Далее выявили общее количество кредитополучателей и общее количество должников
grouped['total_debtors'] = grouped['sum']
grouped['total_clients'] = grouped['count']

# Подсчет доли должников
grouped['debt_ratio'] = (grouped['total_clients'] / grouped['total_debtors']) 

# Удаление лишних столбцов ('count','sum')
grouped.drop(['count', 'sum'], axis=1, inplace=True)

# Переименование индекса в "категория"
grouped.index.name = 'категория'

# Вывод таблицы
print(grouped)


           total_debtors  total_clients  debt_ratio
категория                                          
0                   1063          14149   13.310442
1                    444           4818   10.851351
2                    194           2055   10.592784
3                     27            330   12.222222
4                      4             41   10.250000
5                      0              9         inf


**Вывод:** Из данных усматривается, что с увеличением количества детей, понижается количество должников. Дети явно служат для своих родителей и попечителей как большая ответственность, поэтому минимизируют количество задолженностей в банках. 
    В рамках исключения у нас категория с 3 детьми. Доля должников составляет 12.2%. Для более качественного иследования нам необходимо провести анализ данных по остальным категориям, дабы выявить причину выброса данных по этой категории. 


#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [47]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')

# Произвели группировку данных по категории "количество детей"
grouped = data.groupby('family_status')['debt'].agg(['count', 'sum'])
# Далее выявили общее количество кредитополучателей и общее количество должников
grouped['total_debtors'] = grouped['sum']
grouped['total_clients'] = grouped['count']

# Подсчет доли должников
grouped['debt_ratio'] = (grouped['total_clients'] / grouped['total_debtors']) 

# Удаление лишних столбцов ('count','sum')
grouped.drop(['count', 'sum'], axis=1, inplace=True)

# Переименование индекса в "категория"
grouped.index.name = 'категория'
# Для визуального удобства чтения данных, сортируем категорию по возрастанию. 
grouped_sorted = grouped.sort_values(by=['debt_ratio'])

# Вывод таблицы
print(grouped_sorted)


                       total_debtors  total_clients  debt_ratio
категория                                                      
Не женат / не замужем            274           2813   10.266423
гражданский брак                 388           4177   10.765464
женат / замужем                  931          12380   13.297530
в разводе                         85           1195   14.058824
вдовец / вдова                    63            960   15.238095


**Вывод:**  
Очевидно, что группы населения, находящиеся в тяжелом семейном положении, являются наиболее неплатежеспособными по сравнению с остальными. 
Можно в дальнейшем для исследования взаимосвязь между категориями "в разводе"-"вдовец / вдова" и "количество детей в семье". 

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [53]:
import pandas as pd

# Предположим, что у вас уже есть DataFrame с названием data и столбцами total_income_category и debt

# Группировка данных по категории дохода и подсчет количества записей в каждой категории
category_counts = data['total_income_category'].value_counts().reset_index()
category_counts.columns = ['Категория', 'Количество']

# Группировка данных по категории дохода и подсчет количества должников в каждой категории
debt_counts = data.groupby('total_income_category')['debt'].sum().reset_index()
debt_counts.columns = ['Категория', 'Количество должников']

# Объединение таблиц по категории дохода
merged_data = pd.merge(category_counts, debt_counts, on='Категория')

# Создание дополнительного столбца с процентным соотношением должников в каждой категории от общего числа клиентов в этой категории
merged_data['Процент должников от общего числа'] = (merged_data['Количество должников'] / merged_data['Количество']) * 100

# Вывод объединенной таблицы для сравнения
print(data)


       children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   Среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем        

**Вывод:** 
Поскольку основные клиенты банка-кредитора являются люди с указанным доходом от 50 000 и до 1 000 000, то сравниваем две выборки В и С. Лидирующей категорией среди должников, являются кредитополучатели, указывающие доход от 50 001 до 200 000. 
Наиболее ответственная и массивная часть кредитополучателей является группа с указанным доходом от 200 001 до 1 000 000. 
Остальные выборки необходимо проанализировать по дополнительным критериям (напр. цель получения кредита и наличие детей, поскольку это социально-отягочающие факторы, влияющие на доход).

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [54]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')

# Снова создадим функцию, которая будет объединять категории.
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
        else:
            return 'другое'
    except:
        return 'нет категории'

data['purpose_category'] = data['purpose'].apply(categorize_purpose)

# Создадим сводную таблицу с количеством должников и недолжников для каждой категории
pivot_table = pd.pivot_table(data, index='purpose_category', columns='debt', values='total_income', aggfunc='count')

# Переименовываем столбцы для лучшего понимания
pivot_table.columns = ['no_debt', 'debt']

# Добавляем столбец с процентом должников от общего числа заемщиков для каждой категории целей кредита
pivot_table['debt_percentage'] = (pivot_table['debt'] / (pivot_table['no_debt'] + pivot_table['debt'])) * 100

# Сортируем сводную таблицу по проценту должников
pivot_table_sorted = pivot_table.sort_values(by='debt_percentage', ascending=False)

# Выводим отсортированную сводную таблицу
print(pivot_table_sorted)


                          no_debt  debt  debt_percentage
purpose_category                                        
операции с автомобилем       3530   367         9.417501
получение образования        3266   331         9.202113
проведение свадьбы           1941   158         7.527394
операции с недвижимостью     9043   715         7.327321


**Вывод:** 
Операции с автомобилем имеют самый высокий процент должников. Это может быть связано с тем, что операции с автомобилем часто требуют значительных финансовых затрат, и заемщики могут столкнуться с финансовыми трудностями при возврате кредита.
Получение образования также имеет высокий процент должников в размере 9.20%. Это может быть обусловлено тем, что заемщики, берущие кредит на образование, могут столкнуться с трудностями в погашении кредита из-за длительного времени обучения и невозможности получения стабильного дохода во время учебы.
Данные по первым двум категориям лучше будет проанализировать вместе с gender — полом клиента. 
Проведение свадьбы и операции с недвижимостью имеют более низкий процент должников, составляющий 7.52% и 7.32% соответственно. Это может указывать на то, что эти категории целей кредита обладают более стабильной возвратной способностью, возможно, из-за более четко определенных целей и более высокой финансовой ответственности заемщиков.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
Возможна неполнота сбора данных или человеческие ошибки, а также баги или ошибки в программном обеспечении, которые могут быть вызваны процессами обновления программного обеспечения.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 
Медианное значение устойчиво к выбросам в данных. То есть использование медианы в пропусках помогает сохранить распределение данных и не искажает его форму, особенно в случае наличия выбросов и сохраняет адекватные средние данные.


### Шаг 4: общий вывод.

Из проведенного анализа данных можно сделать следующие выводы:

Влияние количества детей: Не обнаружено существенного влияния количества детей на факт возврата кредита в срок. Это означает, что наличие или отсутствие детей у заемщиков не является ключевым фактором при принятии решений о выдаче кредита.

Влияние семейного положения: Также не обнаружено значительного влияния семейного положения на возврат кредита в срок. Это говорит о том, что заемщики в разных семейных статусах имеют примерно одинаковые шансы на возврат кредита в срок.

Зависимость от уровня дохода: Корреляционный анализ показал, что зависимость между уровнем дохода заемщиков и возвратом кредита в срок практически незначительна. Это означает, что уровень дохода не является решающим фактором при оценке кредитоспособности.

Цели кредитов: Наиболее высокий процент невозврата отмечается в случае операций с автомобилем и получения образования.

Причины появления пропусков: Возможные причины появления пропусков в исходных данных могут быть связаны с неполнотой сбора данных или человеческими ошибками при заполнении анкет.

Рекомендации:

При оценке кредитоспособности заемщиков стоит учитывать не только их семейное положение, количество детей и уровень дохода, но и другие факторы, такие как стабильность работы, кредитная история и срочные финансовые обязательства.
Для минимизации рисков невозврата кредита в срок рекомендуется более тщательно анализировать заемщиков, оценивая их платежеспособность и риски, связанные с конкретными целями кредитования.
Важно систематически собирать и анализировать данные о клиентах, чтобы выявить и корректировать потенциальные проблемные области и оптимизировать процессы выдачи кредитов.
Этот отчет представляет собой свод информации о влиянии различных факторов на возврат кредита в срок и предлагает рекомендации для более эффективного управления кредитным портфелем и минимизации рисков для банка.